0005-R-pubsub-delayed-scheduling.md

```diff
@@ -0,0 +1,156 @@
+
+* What alternatives have been considered, and why do they not solve the problem?
+
+The only alternative for users is to create separate topics for delayed messages together with boilerplate code to orchestrate sending/receiving messages from said topics within their code, and then chaining the message flow to the correct topics after the conditions for sending delayed messages have been met. In addition, user code will need to become aware of Dapr specific mechanisms and mirror it in user code.
+
+* What advantages / disadvantages does this proposal have?
+
+This proposal allows delayed message scheduling to be enabled for:
+
+1. Every pub/sub component supported by Dapr without requiring component specific changes
+2. Messages sent using Dapr CloudEvents
+3. Messages sent using user provided CloudEvents
+4. Messages sent using raw payloads
+5. Message sent/received by HTTP and gRPC subscribers
+
+## Implementation Details
+
+### Design
+
+The Dapr runtime will use an intermediary topic to store messages that are scheduled for late delivery. Users will specify their intent for a message to be delayed using the existing metadata mechanism (specified as a query param on HTTP1.1 or on the request object for gRPC), the same way we use metadata for TTLs in pub/sub and for many other features. The proposed metadata key to be used is `scheduledSendTime` with the value being a date time string adhering to the ISO-8601 format, the same format supported by Dapr actors reminders and timers.
+
+HTTP example:
+
+```bash
+curl -X "POST" http://localhost:3500/v1.0/publish/pubsub/singularity?metadata.scheduledSendTime=2045-11-05T08:15:30-05:00 -H "Content-Type: application/json" -d '{"execute-order": "66"}'
+```
+
+Upon receiving a delayed message, the Dapr runtime will examine the due date for the message and publish it to the target topic if time is due. If the time isn't due, the Dapr runtime will hold the message until the time is right to send it without ACKing back to the broker. If the Dapr instance crashes, the message will remain in the broker and be consumed by a different instance.
+
+#### Configuring delayed messaging
+
+Users can enable delayed messaging by adding the following pub/sub component metadata (other fields specific for Redis removed for brevity):
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: pubsub
+spec:
+  type: pubsub.redis
+  version: v1
+  metadata:
+  - name: "enableDelayedMessaging"
+    value: true
+```
+
+If not set, the default value is `false` and the feature is bypassed completely.
+
+
+#### Consuming delayed messages
+
+On startup, the Dapr runtime will subscribe to a topic in the following format: `<app-id>-delayed`. This topic holds all delayed messages for the given application, and has the following benefits:
+
+1. Topics messages are scoped per application and not per component
```

Comment:
needs clarification:

I'm assuming that some component must be configured at first to be the owner of the `-delayed` topics, correct me if I'm wrong,

so my questions are:

1) How this will be configured?
2) Can the broker send those messages to a dead-letter topic due to the lack of ACKing? If yes, will Dapr subscribe to a dead letter topic while not ACKing the messages?  If not, should we have a way to distinguish between a runtime bug that prevents ACKing messages, and the desired behavior?
3) given that the topic is not component-scoped, what would happen in case of the target pubsub component has been deleted while trying to publish the message ?

Reply:
> How will this be configured?

This is explained below.. see [here](https://github.com/dapr/proposals/pull/12/files#diff-facc607b55c1491e99f88ba1b46ba1c3b3fa8841614b300a6a938239bf3c6d09R59).

> Can the broker send those messages to a dead-letter topic due to the lack of ACKing?

That would be undesirable. Dead letters topics are set per topic and handle dead lettering for the application processing.

> If not, should we have a way to distinguish between a runtime bug that prevents ACKing messages, and the desired behavior?

Delayed messages shouldn't prevent ACKing messages for the end topic, once a delayed message has been published to the end topic it will be ACKed by the app. The runtime will only ACK the message for the delayed message topic after it has been sent to the target topic successfully. Failures to ACK back to the delayed message topic will show up in the logs with the name of the delayed message topic in the same flow we have today.

> given that the topic is not component-scoped, what would happen in case of the target pubsub component has been deleted while trying to publish the message ?

The same thing that happens if a pubsub component is deleted while trying to publish a message to any other topic, be it an app topic, deadletter or otherwise - Dapr returns an error to the user code.

Reply:
> This is explained below.. see [here](https://github.com/dapr/proposals/pull/12/files#diff-facc607b55c1491e99f88ba1b46ba1c3b3fa8841614b300a6a938239bf3c6d09R59).

I was thinking they were different things. One thing is which pubsub will be used to store delayed messages and another thing is whether the pubsub in question supports delayed messages or not. So the `enableDelayedMessaging` has the same semantics as `actorStateStore` for statestore buildingblocks? is that correct?

> That would be undesirable. Dead letters topics are set per user topic and handle dead lettering for the application processing.
Got it,
Just a heads up to make it clear through documentation, I was thinking about specific configurations that might be needed in case of manually creating such `-delayed` topics

> Delayed messages shouldn't prevent ACKing messages for the end topic, once a delayed message has been published to the end topic it will be ACKed by the app.

I'm trying to imagine a scenario where the target pubsub is out for any reason, probably this would lead in n-ACKing messages by dapr runtime, not because its part of the message delay, but because there are something preventing the message to be delivered to the target topic, and, because of this, the message retention would increase indefinitely.

Reply:
> another thing is whether the pubsub in question supports delayed messages or not

Please read the proposal fully. All pub subs support delayed messages.

Reply:
> I was thinking about specific configurations that might be needed in case of manually creating such -delayed topics

There is such a case, explained here: https://github.com/dapr/proposals/pull/12/files#diff-facc607b55c1491e99f88ba1b46ba1c3b3fa8841614b300a6a938239bf3c6d09R118

Reply:
> I'm trying to imagine a scenario where the target pubsub is out for any reason, probably this would lead in n-ACKing messages by dapr runtime, not because its part of the message delay, but because there are something preventing the message to be delivered to the target topic, and, because of this, the message retention would increase indefinitely.

As you mentioned, this is unrelated to delayed scheduling.

Reply:
> Please read the proposal fully. All pub subs support delayed messages

I understand that this functionality is provided by the runtime. My question was if the flag would serve as a feature-flag to enable/disable for a specific pubsub or if it would serve to mark a pubsub as the owner of the `-delayed` topics but you answred it, thanks.

> There is such a case, explained here

I saw that. Just emphasizing that we should address DLQ's specifically

> As you mentioned, this is unrelated to delayed scheduling.

Yup. But it doesn't mean that we can't do anything about it, for example we can have a TTL or a maximum number of retries for attempted delivery, does that make sense?

Reply:
> My question was if the flag would serve as a feature-flag to enable/disable for a specific pubsub or if it would serve to mark a pubsub as the owner of the -delayed topics but you answred it, thanks.

This feature will have a preview feature flag regardless, but there's the specified specific field that explicitly enables delayed messages for a component.

Reply:
> Yup. But it doesn't mean that we can't do anything about it, for example we can have a TTL or a maximum number of retries for attempted delivery, does that make sense?

This kind of overlaps with pub/sub resiliency, but this can and should be discussed in dapr/dapr.

Reply:
Just to correct myself, because I said "delivered" but actually the "delivered" for this message is a "publish" in the target pubsub. 

I understand that the resiliency is more "stateless" than we need for that, but I could be talking nonsense. When I mentioned "retry" I meant number of retries even between different runtimes (in case of a crash for instance) and I don't know how our resiliency policy would currently handle TTL (sorry). But I think the subject in general is about resiliency yes, I just don't know if it would fit into something related to pubsub to be honest.